Source/WebKit/Shared/haiku/WebEventFactory.cpp

```diff
@@ -26,96 +26,88 @@
 #include "config.h"
 #include "WebEventFactory.h"
 
+#include "WebEventModifier.h"
+#include "WebMouseEvent.h"
 #include <WebCore/IntPoint.h>
+#include <wtf/WallTime.h>
 
-namespace WebKit {
+#include <AppDefs.h>
+#include <InterfaceDefs.h>
+#include <Message.h>
+#include <View.h>
 
-enum {
-    BUTTON_PRESS = 'btps',
-    BUTTON_RELEASE = 'btrl',
-    MOUSE_MOVEMENT = 'mmmv',
-}
+namespace WebKit {
+using namespace WebCore;
 
-static inline int clickCount(WebEvent::Type type, WebMouseEvent::Button button, const POINT& position, int64_t timeStampSeconds)
+WebMouseEvent WebEventFactory::createWebMouseEvent(const BMessage* message)
 {
-    static int gLastClickCount;
-    static int64_t gLastClickTime;
-    static POINT lastClickPosition;
-    static WebMouseEvent::Button lastClickButton = WebMouseEvent::LeftButton;
-
-    bool cancelPreviousClick = 0;
-
-    if (type == WebEvent::MouseDown) {
-        if (!cancelPreviousClick && (button == lastClickButton))
-            ++gLastClickCount;
-        else {
-            gLastClickCount = 1;
-            lastClickPosition = position;
-        }
-        gLastClickTime = timeStampSeconds;
-        lastClickButton = button;
-    } else if (type == WebEvent::MouseMove) {
-        if (cancelPreviousClick) {
-            gLastClickCount = 0;
-            lastClickPosition.x = 0;
-            lastClickPosition.y = 0;
-            gLastClickTime = 0;
-        }
+    WebEventType type;
+    switch (message->what) {
+    case B_MOUSE_DOWN:
+        type = WebEventType::MouseDown;
+        break;
+    case B_MOUSE_UP:
+        type = WebEventType::MouseUp;
+        break;
+    case B_MOUSE_MOVED:
+        type = WebEventType::MouseMove;
+        break;
+    default:
+        ASSERT_NOT_REACHED();
     }
 
-    return gLastClickCount;
-}
-
-static inline WebMouseEvent::Button buttonForEvent(const BMessage* message)
-{
-    int32_t buttonType;
-    unsigned button = 0;
-    message->FindInt32("button",&buttonType);
-    switch (buttonType) {
-        case B_PRIMARY_MOUSE_BUTTON:
-            button = WebMouseEvent::LeftButton;
-            break;
-        case B_SECONDARY_MOUSE_BUTTON:
-            button = WebMouseEvent::RightButton;
-            break;
-        case B_TERTIARY_MOUSE_BUTTON:
-            button = WebMouseEvent::MiddleButton;
-            break;
+    WebMouseEventButton button = WebMouseEventButton::None;
+    unsigned short buttons = 0;
+    int32_t nativeButtons;
+    message->FindInt32("buttons", &nativeButtons);
+    // It doesn't appear that the message we receive indicates which mouse
+    // button was pressed to trigger the event, so we can fill in buttons
+    // properly, but button will have to be guessed.
```

Comment:
you don't need to guess, you have to track the state of the buttons from the previous calls to the function, and from that you can tell what changed (for example you can use XOR between the previous and the new state).

Be aware that multiple buttons may change at the same time (this is sometimes used intentionally, for example using both the right and left buttons at the same time will typically simulate a middle click if your mouse has only two buttons)